### ISSUE TITLE
Multiple frontends and one kernel

### ISSUE THREAD

**Issue by maartenbreddels** [jupyter-widgets/ipywidgets#1218]
As requested by @jasongrout I write down what I had in mind.
When working with ipyvolume with the virtual reality / cardboard option, I had the idea that it would be useful to 'expose' a single widget from a running kernel (that is, from the notebook) to a separate 'page'.
Say, to make it more concrete (say in the notebook you run)
```
widget = create_some_widget(...)
somelibrary.expose_widget(widget, 'mywidget')
```
This will return sth like:
```Widget available at url http://localhost:8888/expose/mywidget?token=....```
When you open this url, you'd only see that widget, for instance an ipyvolume widget on the phone for vr rendering. Connected to a live kernel, it would be possible to change the existing widget, for instance change the glyph colors, which will be directly reflected on the frontend running on the phone, and in the notebook.

There are issues with this, and I don't want to go into too much detail, but I think this can now be done already, it is partly already working. However, only changes coming from the backend will be reflected back in all frontends, a change coming from a frontend, will not be reflected in the other frontends.

So maybe when the backend receives an update, it should boardcast this change to all of the frontends (preferrable minus the one it got its changes from, but @jasongrout told me that probably isn't possible right now).

TL;DR: When you have multiple frontends connected to a single backend/kernel, a change in 1 frontend will not be synced to the other frontends.

**Comment by jasongrout** [comment]
when the backend gets a sync change from the frontend, it should probably broadcast that change back out in a special sync_broadcast comm message, similar to how we now broadcast back out any execution input.

**Comment by jasongrout** [comment]
We could manually broadcast the change back out to clients until the protocol is fixed.

**Comment by sjdv1982** [comment]
As a newcomer to ipywidgets, I originally misunderstood this. It's not that the last front-end widget forces itself onto the kernel widget: a many-to-one relation between front-end widgets and kernel widgets is maintained. It's just that if one front-end widget pushes an update, the kernel widget assumes that that all other front-end widgets are already aware of it.
 
For other newcomers who run into this problem, here is a fix: https://github.com/jupyter-widgets/ipywidgets/issues/1881#issuecomment-353696609

**Comment by jasongrout** [comment]
Thank you very much for posting a workaround. That's essentially what I meant above by saying we could manually broadcast the change back out to clients, i.e., we should explore implementing that workaround in the base ipywidgets package automatically.

**Comment by jasongrout** [comment]
(Though it would make sense for us to do the send_state right when we receive state, rather than on a value change)

**Comment by sjdv1982** [comment]
> (Though it would make sense for us to do the send_state right when we receive state, rather than on a value change)
Ah ok. I used a hook-on-value-change because .observe provides exactly this. Does ipywidgets/traitlets have a hook-on-receive-state? If so, I will fix the workaround.

**Comment by jasongrout** [comment]
No, that hook doesn't exist. I think what you did is probably the best workaround until we change the ipywidgets codebase.

Do you want to look at changing the ipywidgets codebase and submitting a PR? The send would probably happen somewhere around https://github.com/jupyter-widgets/ipywidgets/blob/5881cce685e1fb8ebf064735946df377a87e4cbe/ipywidgets/widgets/widget.py#L662 - it may be enough to just send the message we received back to the frontend.

**Comment by sjdv1982** [comment]
I could try a PR (not right now though), but I am not sure... if the payload is rather large, and there is only one front-end, this would be a big waste of bandwidth? So it should be at least configurable?

**Comment by jasongrout** [comment]
That is a problem we've discussed on https://github.com/jupyter/jupyter_client/issues/263.

**Comment by jtpio** [comment]
> if the payload is rather large, and there is only one front-end, this would be a big waste of bandwidth? So it should be at least configurable?

Maybe there could be something similar to `continuous_update` (already used in some core widgets) for this workaround.

**Comment by vidartf** [comment]
There could likely be an configuration/optimization in an implementation of https://github.com/jupyter/jupyter_client/issues/263 to prevent sending echos in all / single-client cases.

**Comment by maartenbreddels** [comment]
implemented in #3195
